src/nbl/video/COpenGL_Swapchain.cpp

```diff
@@ -0,0 +1,414 @@
+                
+            // Create FBO
+            gl->glGeneral.pglEnable(IOpenGL_FunctionTable::FRAMEBUFFER_SRGB);
+            gl->extGlCreateFramebuffers(1, &fbos[request.imgIx]);
+
+            GLuint fbo = fbos[request.imgIx];
+            gl->extGlNamedFramebufferTexture(fbo, GL_COLOR_ATTACHMENT0, m_texViews[request.imgIx], 0, GL_TEXTURE_2D);
+            GLenum drawbuffer0 = GL_COLOR_ATTACHMENT0;
+            gl->extGlNamedFramebufferDrawBuffers(fbo, 1, &drawbuffer0);
+
+            GLenum status = gl->extGlCheckNamedFramebufferStatus(fbo, GL_FRAMEBUFFER);
+            assert(status == GL_FRAMEBUFFER_COMPLETE);
+
+            request.initializeImg = nullptr;
+        }
+        else
+        {
+            needToBlit = false;
+
+            const uint32_t imgix = request.imgIx;
+            const uint32_t w = imgWidth;
+            const uint32_t h = imgHeight;
+
+            for (uint32_t i = 0u; i < request.semCount; ++i)
+            {
+                core::smart_refctd_ptr<COpenGLSemaphore>& sem = request.sems->begin()[i];
+                sem->wait(gl);
+            }
+
+            // need to possibly wait for master context (image & view creation, etc.)
+            m_masterContextCallsWaited = m_device->waitOnMasterContext(gl, m_masterContextCallsWaited);
+
+            gl->extGlBlitNamedFramebuffer(fbos[imgix], 0, 0, 0, w, h, 0, 0, w, h, GL_COLOR_BUFFER_BIT, GL_NEAREST);
+            syncs[imgix] = core::make_smart_refctd_ptr<COpenGLSync>();
+            syncs[imgix]->init(core::smart_refctd_ptr<IOpenGL_LogicalDevice>(m_device), gl, false);
+            // swap buffers performs an implicit flush before swapping 
+            // https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglSwapBuffers.xhtml
+            egl->call.peglSwapBuffers(egl->display, glctx.surface);
+        }
+    }
+
+    void exit(SThreadHandlerInternalState* state)
+    {
+        // TODO this should be in destructor for COpenGLImage
+        auto gl = getFunctionPointer(m_device->getAPIType(), state);
+        gl->~IOpenGL_FunctionTable();
+
+        egl->call.peglMakeCurrent(egl->display, EGL_NO_SURFACE, EGL_NO_SURFACE, EGL_NO_CONTEXT);
+        egl->call.peglDestroyContext(egl->display, glctx.ctx);
+        egl->call.peglDestroySurface(egl->display, glctx.surface);
+    }
+
+    bool wakeupPredicate() const { return needToBlit; }
+    bool continuePredicate() const { return needToBlit; }
+
+private:
+    core::smart_refctd_ptr<IOpenGL_LogicalDevice> m_device;
+    uint64_t m_masterContextCallsWaited;
+
+    const egl::CEGL* egl;
+    ISurface::E_PRESENT_MODE m_presentMode;
+    const COpenGLFeatureMap* features;
+    uint32_t imageCount;
+    uint32_t imgWidth, imgHeight;
+    uint32_t fbos[MaxImages]{};
+    core::smart_refctd_ptr<COpenGLSync> syncs[MaxImages];
+    COpenGLDebugCallback* m_dbgCb;
+    std::array<uint32_t, MaxImages> m_texViews;
+
+    bool needToBlit = false;
```

Comment:
this needs to be an atomic flag, otherwise you end up with memory ordering issues (lack of barriers, out of order execution, etc)